dds/DCPS/XTypes/DynamicDataImpl.cpp

```diff
@@ -1472,6 +1472,19 @@ DDS::ReturnCode_t DynamicDataImpl::set_boolean_value(DDS::MemberId id, CORBA::Bo
 
 DDS::ReturnCode_t DynamicDataImpl::set_string_value(DDS::MemberId id, const char* value)
 {
+  DDS::DynamicType_var mtype;
+  DDS::ReturnCode_t rc = get_member_type(mtype, type_, id);
+  if (rc != DDS::RETCODE_OK) {
+    return rc;
+  }
+  if (mtype->get_kind() == TK_ENUM) {
+    DDS::Int32 intValue;
+    rc = get_enumerator_value(intValue, value, mtype);
+    if (rc != DDS::RETCODE_OK) {
+      return rc;
+    }
+    return set_single_value<TK_INT32>(id, intValue);
```

Comment:
I think this needs to call `set_single_value` with a `TypeKind` template argument depending on the bit bound of the enum because it will check the representing integer type of the enum against the template argument. For example, if the enum has a bit bound in [1, 8], it will be represented by a `TK_INT8` integer and this call will fail.

Reply:
Right, I think this can just go through one of the helper functions that Fred provided.